src/main/kotlin/org/flaxo/plagiarism/element/Arrow.kt

```diff
@@ -0,0 +1,116 @@
+
+    /**
+     * Updates coordinates of all required components of the arrow.
+     */
+    fun update(sourceNode: ForceNode, targetNode: ForceNode, directed: Boolean) {
+        with(line) {
+            x1 = sourceNode.x
+            y1 = sourceNode.y
+            x2 = targetNode.x
+            y2 = targetNode.y
+        }
+
+        if (directed && link.directedTo != null) updateTails(link.directedTo)
+    }
+
+    private fun updateTails(direction: Direction) {
+        val dx = line.run { x2 - x1 }
+        val dy = line.run { y2 - y1 }
+        val len = sqrt(dx * dx + dy * dy)
+        val sin = dy / len
+        val cos = dx / len
+
+        // Find actual length of the tails, reduce it if the line is too small
+        val adaptedArrowLengthX = if (Configuration.Link.tailLengthX > len / Configuration.Link.maxTailLengthProportionX)
+            len / Configuration.Link.maxTailLengthProportionX else Configuration.Link.tailLengthX
+        val adaptedArrowLengthY = if (Configuration.Link.tailLengthY > len / Configuration.Link.maxTailLengthProportionY)
+            len / Configuration.Link.maxTailLengthProportionY else Configuration.Link.tailLengthY
+
+        val arrowLengthX: Double
+        val arrowLengthY: Double
+        val arrowHeadX: Double
+        val arrowHeadY: Double
+
+        // Define the coordinates of the tails
+        when (direction) {
+            Direction.FIRST -> {
+                arrowLengthX = adaptedArrowLengthX
+                arrowLengthY = adaptedArrowLengthY
+                arrowHeadX = line.x1
+                arrowHeadY = this.line.y1
+            }
+            Direction.SECOND -> {
+                arrowLengthX = -adaptedArrowLengthX
+                arrowLengthY = adaptedArrowLengthY
+                arrowHeadX = line.x2
+                arrowHeadY = line.y2
+            }
+        }
+
+        with(tails.first) {
+            x1 = calculateRelativeX(arrowLengthX, arrowLengthY, sin, cos, arrowHeadX)
+            y1 = calculateRelativeY(arrowLengthX, arrowLengthY, sin, cos, arrowHeadY)
+            x2 = arrowHeadX
+            y2 = arrowHeadY
+        }
+
+        with(tails.second) {
+            x1 = calculateRelativeX(arrowLengthX, -arrowLengthY, sin, cos, arrowHeadX)
+            y1 = calculateRelativeY(arrowLengthX, -arrowLengthY, sin, cos, arrowHeadY)
+            x2 = arrowHeadX
+            y2 = arrowHeadY
+        }
+    }
+
+
+    /**
+     * Calculate relative x coordinate by the position of [x0]
+     */
+    private fun calculateRelativeX(x: Double, y: Double, sin: Double, cos: Double, x0: Double) =
+            x * cos - y * sin + x0
+
+    /**
+     * Calculate relative y coordinate by the position of [y0]
+     */
+    private fun calculateRelativeY(x: Double, y: Double, sin: Double, cos: Double, y0: Double) =
+            x * sin + y * cos + y0
+
+
+    private fun visibleComponents(directed: Boolean) = if (directed) listOf(line) + tails.toList() else listOf(line)
```

Comment:
You call this function in the `hide` function as `visibleComponents(true)`. I think it should be named like `arrowComponents(withTails: Boolean)`.